File: RESTFul/IMMN/PHP/app1/lib/IMMN/Delta.php
<?php
namespace Att\Api\IMMN;

/*
 * Copyright 2014 AT&T
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

require_once __DIR__ . '/DeltaChange.php';

use Att\Api\IMMN\IMMNDeltaChange;

/**
 * Immutable class used to hold a Delta object.
 *
 * @category API
 * @package  IMMN
 * @author   pk9069
 * @license  http://www.apache.org/licenses/LICENSE-2.0
 * @version  Release: @package_version@ 
 */
final class Delta
{
    private $_type;
    private $_adds;
    private $_deletes;
    private $_updates;

    public function __construct($type, $adds, $deletes, $updates)
    {
        $this->_type = $type;
        $this->_adds = $adds;
        $this->_deletes = $deletes;
        $this->_updates = $updates;
    }

    public function getDeltaType()
    {
        return $this->_type;
    }

    public function getAdds()
    {
        return $this->_adds;
    }

    public function getDeletes()
    {
        return $this->_deletes;
    }

    public function getUpdates()
    {
        return $this->_updates;
    }

    private static function getDeltaChanges($arr)
    {
        $deltaChanges = array();
        foreach($arr as $deltaChangeArr) {
            $deltaChange = IMMNDeltaChange::fromArray($deltaChangeArr);
            $deltaChanges[] = $deltaChange;
        }

        return $deltaChanges;
    }

    public static function fromArray($arr)
    {
        $type = $arr['type'];
        $adds = Delta::getDeltaChanges($arr['adds']); 
        $deletes = Delta::getDeltaChanges($arr['deletes']); 
        $updates = Delta::getDeltaChanges($arr['updates']); 

        return new Delta($type, $adds, $deletes, $updates);
    }
    
}

/* vim: set expandtab tabstop=4 shiftwidth=4 softtabstop=4: */
?>

File: RESTFul/IMMN/Java/app1/src/main/java/com/att/api/immn/service/Delta.java
package com.att.api.immn.service;

import org.json.JSONArray;
import org.json.JSONObject;

public final class Delta {
    private final String type;
    private final DeltaChange[] adds;
    private final DeltaChange[] deletes;
    private final DeltaChange[] updates;

    public Delta(String type, DeltaChange[] adds, DeltaChange[] deletes,
            DeltaChange[] updates) {
            
        this.type = type;
        this.adds = adds;
        this.deletes = deletes;
        this.updates = updates;
    }

    public String getType() {
        return type;
    }

    public DeltaChange[] getAdds() {
        return adds;
    }

    public DeltaChange[] getDeletes() {
        return deletes;
    }

    public DeltaChange[] getUpdates() {
        return updates;
    }

    public static Delta valueOf(JSONObject jobj) {
        String type = jobj.getString("type");

        JSONArray jadds = jobj.getJSONArray("adds");
        DeltaChange[] adds = new DeltaChange[jadds.length()];
        for (int i = 0; i < jadds.length(); ++i) {
            JSONObject jchange = jadds.getJSONObject(i);
            adds[i] = DeltaChange.valueOf(jchange);
        }

        JSONArray jdeletes = jobj.getJSONArray("deletes");
        DeltaChange[] deletes = new DeltaChange[jdeletes.length()];
        for (int i = 0; i < jdeletes.length(); ++i) {
            JSONObject jchange = jdeletes.getJSONObject(i);
            deletes[i] = DeltaChange.valueOf(jchange);
        }

        JSONArray jupdates = jobj.getJSONArray("updates");
        DeltaChange[] updates = new DeltaChange[jupdates.length()];
        for (int i = 0; i < jupdates.length(); ++i) {
            JSONObject jchange = jupdates.getJSONObject(i);
            updates[i] = DeltaChange.valueOf(jchange);
        }

        return new Delta(type, adds, deletes, updates);
    }

}
